### PR TITLE
[SourceKit/CodeFormat] Indent lines in multi-line strings

### PR DESCRIPTION
**Pull request by bnbarham**
The first commit fixes a couple bugs in the Lexer::getLocFor*Line methods in a couple edge cases.

The second updates the formatter to indent multi-line strings where it can. When it's an un-closed string it will indent new lines to the first previous line that isn't whitespace-only, unless that line is the same as the start quotes. In that case it will either indent to the start quotes if that line is only quotes or add an extra indent level otherwise. For closed strings we have to be a little more careful since we don't want to add significant (and possibly unintended) whitespace, so those lines are only indented when they would otherwise be invalid (ie. less than the end quotes).

This could add trailing whitespace to existing lines, but I think that's better behaviour than what's there now. This could be fixed if the whole range was passed in instead of a line at a time, as the indentation could be different when it is *actually* a single line in that case.

Resolves rdar://32181422

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by nathawes — APPROVED** [review]
Looks good as is, but thought we could maybe still handle the target line has content but there are no end quotes case if you think that makes sense too.

File: lib/IDE/Formatting.cpp [hunk 1]
```diff
@@ -1234,14 +1252,94 @@ class FormatWalker : public ASTWalker {
...
+        getLocForContentStartOnSameLine(SM, StringLiteralRange.getEnd());
+    bool HaveEndQuotes = CharSourceRange(SM, EndLineContentLoc,
+                                         StringLiteralRange.getEnd())
+        .str().equals(StringRef("\"\"\""));
+
+    bool TargetLineEmpty = isLineAtLocEmpty(SM, TargetLineLoc);
+    if (!TargetLineEmpty && !HaveEndQuotes) {
+      // Target line has contents but there are no end quotes - don't change
+      // indentation at all.
+      return IndentContext {TargetLocation, false, IndentContext::Exact};
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply lib/IDE/Formatting.cpp]: Yeah, sounds reasonable - changed it to follow the same pattern as if the line was empty.

**Comment by bnbarham** [comment]
@swift-ci please test

File: lib/IDE/Formatting.cpp [hunk 2]
```diff
@@ -1234,14 +1256,94 @@ class FormatWalker : public ASTWalker {
     if (InnermostCtx)
       CtxOverride.applyIfNeeded(SM, *InnermostCtx);
 
+    if (StringLiteralRange.isValid())
+      InnermostCtx = indentWithinStringLiteral();
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply lib/IDE/Formatting.cpp:1258]: Added to this condition

**Comment by nathawes** [comment]
@swift-ci please test and merge

**Comment by nathawes** [comment]
@swift-ci please test and merge

**Pull request merged**